typescript/db/src/repository/workspace.ts

```diff
@@ -47,73 +54,99 @@ export const getWorkspace: Repository["workspace"]["get"] = async (
+    where,
+  });
+  return addTypeToWorkspace(workspaceFromDb);
 };
 
 export const listWorkspace: Repository["workspace"]["list"] = async (
   where,
   skip = undefined,
   first = undefined
 ) => {
-  if (where?.user?.id == null) {
-    return [];
-  }
-  const workspacesFromDb = await (
-    await getPrismaClient()
-  ).workspace.findMany(
-    castObjectNullsToUndefined({
-      skip: skip ?? undefined,
-      take: first ?? undefined,
-      orderBy: { createdAt: Prisma.SortOrder.asc },
-      where: {
-        memberships: { some: { userId: where?.user?.id } },
-        slug: where?.slug ?? undefined,
-      },
-    })
-  );
+  if (isNil(where?.user?.id)) return [];
+  const db = await getPrismaClient();
+  const workspacesFromDb = await db.workspace.findMany({
+    skip: skip ?? undefined,
+    take: first ?? undefined,
+    orderBy: { createdAt: Prisma.SortOrder.asc },
+    where: {
+      memberships: { some: { userId: where?.user?.id ?? undefined } },
+      slug: where?.slug ?? undefined,
+      // Don't list deleted workspaces
+      deletedAt: { equals: null },
+    },
+  });
   return workspacesFromDb.map(addTypeToWorkspace);
 };
 
+/**
+ * Throws an error if the given where filter has a deletedAt property
+ */
+const noDeleteUpdate = (
+  workspace: PartialWithNullAllowed<DbWorkspaceWithType>
+) => {
+  const hasDeletedAt = Object.keys(workspace).includes("deletedAt");
+  if (hasDeletedAt) {
+    throw new Error("Cannot update internal property deletedAt");
+  }
+};
+
 export const updateWorkspace: Repository["workspace"]["update"] = async (
   where,
   workspace,
   user
 ) => {
-  // Check if user has access to workspace, this will throw it it does not
+  noDeleteUpdate(workspace);
   await checkUserAccessToWorkspace({ user, where });
-  // Update workspace
   const newNameAndSlugs =
     typeof workspace.name === "string"
       ? {
           name: workspace.name,
-          slug: slugify(workspace.name, { lower: true }),
+          slug: getSlug(workspace.name),
         }
       : // needed to make prisma happy with the types
         { name: undefined };
   try {
-    await (
-      await getPrismaClient()
-    ).workspace.update({
+    const db = await getPrismaClient();
+    await db.workspace.update({
       where: castObjectNullsToUndefined(where),
       data: castObjectNullsToUndefined({
         ...workspace,
         ...newNameAndSlugs,
+        deletedAt: undefined,
       }),
     });
     return true;
   } catch (e) {
     return false;
   }
 };
+
+export const deleteWorkspace: Repository["workspace"]["delete"] = async (
+  where,
+  user
+) => {
+  const db = await getPrismaClient();
+  const data = await getWorkspace(where, user);
+  if (isNil(data)) {
+    throw new Error("Could not find workspace");
+  }
+  const { stripeCustomerId, slug } = data;
+  await db.workspace.update({
+    where: castObjectNullsToUndefined(where),
+    data: { deletedAt: new Date().toISOString(), slug: `${slug}-${data.id}` },
```

Comment:
You also need to change the `name` of the workspace, or you'll end up having conflicting name issues (see general comment)